tools/wpt/revlist.py

```diff
@@ -0,0 +1,205 @@
+    },
+    "epoch_step": {
+        "name": "epoch-step",
+        "default": DEFAULT_EPOCH_STEP,
+        "help": "Regular interval of seconds used to get the tagged revision. "
+                "hourly, two-hourly, four-hourly, six-hourly, eight-hourly, "
+                "twelve-hourly, daily, weekly or 3600, 7200, 86000 ...",
+        "parser": parser_epoch_step
+    },
+    "num_revisions": {
+        "name": "num-revisions",
+        "default": DEFAULT_NUM_REVISIONS,
+        "help": "Number of revisions to get.",
+        "parser": parser_num_revisions
+    },
+    "log_level": {
+        "name": "log-level",
+        "default": DEFAULT_LOG_LEVEL,
+        "help": "Debug log level. For example: DEBUG, INFO, WARNING, ERROR, "
+                "CRITICAL or an integer value like 10, 20 ",
+        "parser": parser_debug_log_level
+    },
+    "verbose": {
+        "name": "verbose",
+        "default": DEFAULT_VERBOSE,
+        "help": "Show more information in the output",
+        "parser": parser_verbose,
+        "type": bool
+    }
+}
+
+
+def get_git_cmd(repo_path):
+    # type: (bytes) -> Callable[..., Text]
+    """Create a function for invoking git commands as a subprocess."""
+    def git(cmd, *args):
+        # type: (Text, *Union[bytes, Text]) -> Text
+        full_cmd = [u"git", cmd] + list(item.decode("utf8") if isinstance(item, bytes) else item for item in args)  # type: List[Text]
+        try:
+            logger.debug(" ".join(full_cmd))
+            return subprocess.check_output(full_cmd, cwd=repo_path).decode("utf8").strip()
+        except subprocess.CalledProcessError as e:
+            logger.critical("Git command exited with status %i" % e.returncode)
+            logger.critical(e.output)
+            sys.exit(1)
+    return git
+
+
+def list_tagged_revisons(**kwargs):
+    # type: (**Any) -> List[Text]
+    logger.debug("list_tagged_revisons: %s" % kwargs)
+    git = get_git_cmd(wpt_root)
+    epoch_step = COMMAND_ARGS["epoch_step"]["parser"](**kwargs)
+    epoch_until = COMMAND_ARGS["epoch_until"]["parser"](**kwargs)
+    num_revisions = COMMAND_ARGS["num_revisions"]["parser"](**kwargs)
+    verbose = COMMAND_ARGS["verbose"]["parser"](**kwargs)
+
+    iterator = 0
+    result = []
+    previous_command_output = ""
+    while iterator < num_revisions:
+        min_age = (( epoch_until / epoch_step ) - iterator) * epoch_step
+        # For epoch size equal to weekly we need to use a offset to start
+        # the week on Monday 00:00:00.
+        if epoch_step == WEEKLY_EPOCH_SIZE:
+            min_age += WEEKLY_EPOCH_OFFSET
+        if min_age < 0:
+            break
+        command = ["--max-count=1", "--tags=%s" % TAGGED_REVISIONS,
+                   "--min-age=%s" % min_age]
+        if verbose:
+            command.append("--format=date: %cI epoch: %ct commit: %H %D")
+        command.append("--no-walk")
+        command.append(HEAD_REVISION)
+        command_output = git("rev-list", *command)
```

Comment:
`./wpt rev-list --epoch-step 3600 --num-revisions 100` takes about 12 second to run locally on my very powerful workstation, and I suspect calling `git rev-list` 100 times is part of the reason, it effectively makes this an O(n^2) algorithm instead of O(n).

Calling `git for-each-ref 'refs/tags/merge_pr_*' --sort=committerdate --format="%(refname:lstrip=2) %(committerdate:iso-strict)"` takes about 0.3 seconds even though it prints all the information we need, so that's a baseline. Not all of that output will be needed but I think it's fine to read all candidates until that begins taking too much time, in a few years perhaps.

To get the commit hashes instead of the tag names you can use `--format="%(objectname) %(committerdate:iso-strict)"`.

To correctly deal with origin/master not yet being tagged, `refs/remotes/origin/master` can be added to the command line. However, may I suggest starting with only considering the tagged commits as candidates and then in step two see if the most natural place to add it is in this script or the cron action that calls it?

Reply:
Note that if `refs/remotes/origin/master` is added and it is already a tagged commit, then the commit will be listed twice. Easy to deal with, but as I said I'd suggest doing it separately.

Reply:
Yeah, I didn't take care about this. I also could consider to iterate the `git for-each-ref`  with `max-count` and `skip` to avoid  request for all the MR tags at the same time.

Reply:
I made some changes. Now the command iters the rev-list command in bunchs of `10000` revisions:

```
$ reset;  time ./wpt rev-list --num-revisions 100 --epoch-step weekly --epoch-until now  --log-level 10 --verbose --head upstream/master
DEBUG:root:Log level set to 10
DEBUG:root:list_tagged_revisons: {'log_level': '10', 'verbose': True, 'num_revisions': '100', 'head_revision': 'upstream/master', 'epoch_until': 'now', 'epoch_step': 'weekly'}
DEBUG:root:git rev-list --no-walk --skip=0 --max-count=10000 --min-age=1570522963 --format=epoch:%ct:commit:%H:%D upstream/master
DEBUG:root:git rev-list --no-walk --skip=10000 --max-count=10000 --min-age=1570522963 --format=epoch:%ct:commit:%H:%D upstream/master
epoch: 1570403055 commit: 080acb0967ec06504808de9a25f64290df3d6128 tag: merge_pr_19530, upstream/epochs/weekly, upstream/epochs/daily
...
epoch: 1510493090 commit: 37d7c85e7c87000482073c609be3cfdce43a6452 tag: merge_pr_8158

real	0m0,842s
user	0m0,772s
sys	0m0,122s
```

* In execution time is still `O(n^2)`  but with a `n`  so low (each n are process 10000 items)
* In memory is still `O(n)` where `n`  is 10000 items.